flow/device_flow.go

```diff
@@ -0,0 +1,205 @@
+	HandledAt sqlxx.NullTime      `db:"handled_at"`
+	Error     *RequestDeniedError `db:"error"`
+}
+
+// NewDeviceFlow return a new DeviceFlow from a DeviceUserAuthRequest.
+func NewDeviceFlow(r *DeviceUserAuthRequest) *DeviceFlow {
+	return &DeviceFlow{
+		ID:          r.ID,
+		Client:      r.Client,
+		ClientID:    r.ClientID,
+		RequestURL:  r.RequestURL,
+		Verifier:    r.Verifier,
+		CSRF:        r.CSRF,
+		RequestedAt: r.RequestedAt,
+		WasHandled:  r.WasHandled,
+		HandledAt:   r.HandledAt,
+		State:       DeviceFlowStateInitialized,
+	}
+}
+
+// GetDeviceUserAuthRequest return the DeviceUserAuthRequest from a DeviceFlow.
+func (f *DeviceFlow) GetDeviceUserAuthRequest() *DeviceUserAuthRequest {
+	return &DeviceUserAuthRequest{
+		ID:          f.ID,
+		Client:      f.Client,
+		ClientID:    f.ClientID,
+		RequestURL:  f.RequestURL,
+		Verifier:    f.Verifier,
+		CSRF:        f.CSRF,
+		RequestedAt: f.RequestedAt,
+		WasHandled:  f.WasHandled,
+		HandledAt:   f.HandledAt,
+	}
+}
+
+// GetHandledDeviceUserAuthRequest return the HandledDeviceUserAuthRequest from a DeviceFlow.
+func (f *DeviceFlow) GetHandledDeviceUserAuthRequest() *HandledDeviceUserAuthRequest {
+	return &HandledDeviceUserAuthRequest{
+		ID:                f.ID,
+		Client:            f.Client,
+		ClientID:          f.ClientID,
+		Request:           f.GetDeviceUserAuthRequest(),
+		RequestURL:        f.RequestURL,
+		RequestedAt:       f.RequestedAt,
+		RequestedScope:    f.RequestedScope,
+		RequestedAudience: f.RequestedAudience,
+		WasHandled:        f.WasHandled,
+		HandledAt:         f.HandledAt,
+		Error:             f.Error,
+	}
+}
+
+// HandleDeviceUserAuthRequest updates the flows fields from a handled request.
+func (f *DeviceFlow) HandleDeviceUserAuthRequest(h *HandledDeviceUserAuthRequest) error {
+	if f.WasHandled {
+		return errors.WithStack(x.ErrConflict.WithHint("The user_code was already used and can no longer be changed."))
+	}
+
+	if f.State != DeviceFlowStateInitialized && f.State != DeviceFlowStateUnused && f.State != DeviceFlowStateError {
+		return errors.Errorf("invalid flow state: expected %d/%d/%d, got %d", DeviceFlowStateInitialized, DeviceFlowStateUnused, DeviceFlowStateError, f.State)
+	}
+
+	if f.ID != h.ID {
+		return errors.Errorf("flow device challenge ID %s does not match HandledDeviceUserAuthRequest ID %s", f.ID, h.ID)
+	}
+
+	f.State = DeviceFlowStateUnused
+	if h.Error != nil {
+		f.State = DeviceFlowStateError
+	}
+	f.Client = h.Client
+	f.ClientID = h.ClientID
+	f.HandledAt = h.HandledAt
+	f.WasHandled = h.WasHandled
+	f.Error = h.Error
+
+	return nil
+}
+
+// InvalidateDeviceRequest shifts the flow state to DeviceFlowStateUsed. This
+// transition is executed upon device completion.
+func (f *DeviceFlow) InvalidateDeviceRequest() error {
+	if f.State != DeviceFlowStateUnused && f.State != DeviceFlowStateError {
+		return errors.Errorf("invalid flow state: expected %d or %d, got %d", DeviceFlowStateUnused, DeviceFlowStateError, f.State)
```

Comment:
nitpick: how about we map the state integer to a more explicit string for the error?

Reply:
Fair, but given that we will refactor this later on (to merge it with the flow struct) I think it's not worth the effort to add right now.